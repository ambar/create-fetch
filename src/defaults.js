import toHeaders from './utils/toHeaders'
import toObject from './utils/toObject'

export default (defaults = {}) => fetch => (url, options = {}) =>
  fetch(url, {
    ...defaults,
    ...options,
    headers: toObject([
      // same-key values will be deduped
      ...toHeaders(defaults.headers),
      ...toHeaders(options.headers),
    ]),
  })
