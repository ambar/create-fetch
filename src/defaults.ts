import toHeaders from './utils/toHeaders'
import toObject from './utils/toObject'
import {FetchEnhancer} from './types'

export default (defaults: RequestInit = {}): FetchEnhancer =>
  (fetch) =>
  (url, options = {}) =>
    fetch(url, {
      ...defaults,
      ...options,
      headers: toObject([
        // same-key values will be deduped
        ...toHeaders(defaults.headers),
        ...toHeaders(options.headers),
      ]),
    })
