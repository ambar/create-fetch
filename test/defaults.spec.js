import defaults from '../src/defaults'

describe('defaults', () => {
  const fetch = (...args) => Promise.resolve(args)

  it('should add defaults', async () => {
    const myFetch = defaults({credentials: 'include'})(fetch)
    expect(await myFetch('/')).toMatchSnapshot()
  })

  it('should add defaults and headers', async () => {
    const myFetch = defaults({
      credentials: 'include',
      headers: {
        'content-type': 'text/plain',
      },
    })(fetch)
    // add
    expect(
      await myFetch('/', {headers: {'x-version': '1.0'}})
    ).toMatchSnapshot()
    // overwrite
    expect(
      await myFetch('/', {
        credentials: 'omit',
        headers: {'content-type': 'text/json'},
      })
    ).toMatchSnapshot()
  })

  it('should work without options', async () => {
    const myFetch = defaults()(fetch)
    expect(await myFetch('/')).toMatchSnapshot()
  })

  it('should accept Headers', async () => {
    const headers = new Headers({a: 1, b: 1})
    headers.append('a', 2)
    const myFetch = defaults({headers})(fetch)
    const result = await myFetch('/api', {headers: {b: 2, c: 1}})
    expect(result).toEqual(
      await myFetch('/api', {headers: {a: '1, 2', b: '2', c: '1'}})
    )
    expect(result).toEqual(
      await myFetch('/api', {headers: new Headers({a: '1, 2', b: '2', c: '1'})})
    )
  })
})
