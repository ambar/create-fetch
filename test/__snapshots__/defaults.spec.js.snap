// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaults should accept Headers 1`] = `
Array [
  "/",
  Object {
    "headers": Object {
      "a": "1, 2",
      "b": "1",
      "c": "1",
    },
  },
]
`;

exports[`defaults should accept Headers 2`] = `
Array [
  "/",
  Object {
    "headers": Object {
      "a": "1, 2",
      "b": "2",
      "c": "1",
    },
  },
]
`;

exports[`defaults should add defaults 1`] = `
Array [
  "/",
  Object {
    "credentials": "include",
    "headers": Object {},
  },
]
`;

exports[`defaults should add defaults and headers 1`] = `
Array [
  "/",
  Object {
    "credentials": "include",
    "headers": Object {
      "content-type": "text/plain",
      "x-version": "1.0",
    },
  },
]
`;

exports[`defaults should add defaults and headers 2`] = `
Array [
  "/",
  Object {
    "credentials": "omit",
    "headers": Object {
      "content-type": "text/json",
    },
  },
]
`;

exports[`defaults should filter null or undefined 1`] = `
Array [
  "/",
  Object {
    "headers": Object {
      "c": "0",
      "d": "false",
      "e": "",
      "g": "1",
    },
  },
]
`;

exports[`defaults should work without options 1`] = `
Array [
  "/",
  Object {
    "headers": Object {},
  },
]
`;
